pallets/operations/src/impls.rs

```diff
@@ -1,67 +1,85 @@
 #![allow(clippy::nonminimal_bool)]
 
-use frame_support::{
-    dispatch::DispatchResultWithPostInfo, pallet_prelude::Get, traits::LockIdentifier,
-    WeakBoundedVec,
-};
+use frame_support::{dispatch::DispatchResult, traits::LockIdentifier, WeakBoundedVec};
 use pallet_balances::BalanceLock;
 use parity_scale_codec::Encode;
 use sp_core::hexdisplay::HexDisplay;
 use sp_runtime::DispatchError;
 
 use crate::{
     pallet::{Config, Event, Pallet},
-    traits::{AccountInfoProvider, BalancesProvider, NextKeysSessionProvider},
+    traits::{AccountInfoProvider, BalancesProvider, BondedStashProvider, NextKeysSessionProvider},
     LOG_TARGET, STAKING_ID, VESTING_ID,
 };
 
 impl<T: Config> Pallet<T> {
     /// Checks if account has an underflow of `consumers` counter. In such case, it increments
     /// it by one.
-    pub fn fix_underflow_consumer_counter(who: T::AccountId) -> DispatchResultWithPostInfo {
-        let mut weight = T::DbWeight::get().reads(1);
-        let consumers = T::AccountInfoProvider::get_consumers(&who);
+    pub fn fix_underflow_consumer_counter(who: T::AccountId) -> DispatchResult {
+        let current_consumers = T::AccountInfoProvider::get_consumers(&who);
+        let mut expected_consumers: u32 = 0;
 
-        weight += T::DbWeight::get().reads(1);
-        if Self::no_consumers_some_reserved(&who, consumers) {
-            Self::increment_consumers(who)?;
-            weight += T::DbWeight::get().writes(1);
-            return Ok(Some(weight).into());
+        if Self::reserved_or_frozen_non_zero(&who) {
+            expected_consumers += 1;
+        }
+        let has_vesting_lock = Self::has_vesting_lock(&who);
+        let has_staking_lock = Self::has_staking_lock(&who);
+        if has_staking_lock || has_vesting_lock {
+            expected_consumers += 1;
+            if has_staking_lock {
+                expected_consumers += 1;
+            }
+        }
```

Comment:
Why is the above like that, and not:
```
if has_staking_lock { expected_consumers += 1;}
if has_vesting_lock { expected_consumers += 1;}
```

Reply:
1. The `consumers` counter logic regarding locks is as [follows](https://github.com/Cardinal-Cryptography/polkadot-sdk/blob/aleph-v1.2.0/substrate/frame/balances/src/lib.rs#L1004-L1023) (in version 13):
* when there are no locks and one is added, `consumers` is increased
* when the last lock is removed, `consumers` is decreased.
Vesting pallet as itself does not increase `consumers` counter, only [sets a vesting lock](https://github.com/Cardinal-Cryptography/polkadot-sdk/blob/aleph-v1.2.0/substrate/frame/vesting/src/lib.rs#L598).
Note: this locking `consumers` logic is removed in `polkadot-sdk` 1.4.0 by this change https://github.com/paritytech/polkadot-sdk/pull/1976

2. In addition to 1., `staking.bond` [creates additional lock, always on stash](https://github.com/paritytech/polkadot-sdk/pull/1976). It also [sets staking lock, always on stash account](https://github.com/Cardinal-Cryptography/polkadot-sdk/blob/aleph-v1.2.0/substrate/frame/staking/src/pallet/impls.rs#L285), but it falls into increment described in 1.